### PR TITLE
CP-15132: Systemd conversion

### PR DESCRIPTION
**Pull request by rafalmiel**

### CODE REVIEW

**Comment by robhoes** [comment]
Please squash the reverted reverted commit into the appropriate patch(es). It has changed a lot, so there is no point in keeping it.

**Comment by robhoes** [comment]
There are merge conflicts now.

**Comment by robhoes** [comment]
As discussed in person, please change the type of the systemd service to "simple".

**Comment by robhoes** [comment]
As discussed, please add a systemd target for xe-toolstack-restart, rather than redefining the startup order in the script.

**Comment by robhoes** [comment]
Instead of `init.d-xapi-old`, `init.d-xapi-compat` would be a better name. Also, do you know what actually needs it?

File: scripts/OMakefile [hunk 1]
```diff
@@ -44,14 +44,19 @@ install:
...
-	$(IPROG) init.d-xapi-domains $(DESTDIR)/etc/rc.d/init.d/xapi-domains
-	$(IPROG) init.d-management-interface $(DESTDIR)/etc/rc.d/init.d/management-interface
-	$(IPROG) init.d-genptoken $(DESTDIR)/etc/rc.d/init.d/genptoken
-	$(IPROG) init.d-save-boot-info $(DESTDIR)/etc/rc.d/init.d/save-boot-info
-	$(IPROG) init.d-attach-static-vdis $(DESTDIR)/etc/rc.d/init.d/attach-static-vdis
-	$(IDATA) cdrommon@.service $(DESTDIR)/usr/lib/systemd/system/cdrommon@.service
+	$(IPROG) init.d-management-interface $(DESTDIR)$(LIBEXECDIR)/legacy-management-interface
+	$(IPROG) init.d-save-boot-info $(DESTDIR)$(LIBEXECDIR)/save-boot-info
+	$(IPROG) init.d-attach-static-vdis $(DESTDIR)$(LIBEXECDIR)/attach-static-vdis
+	$(IPROG) cdrommon@.service $(DESTDIR)/usr/lib/systemd/system/cdrommon@.service
```
Comment (robhoes) [inline scripts/OMakefile]: Do these service files need to be executable (`$(IPROG)`)?

**Comment by rafalmiel** [comment]
init.d-xapi-old Is only needed by xenrt testcase when the testcase is fixed we will be able to get rid of it.

File: scripts/OMakefile [hunk 2]
```diff
@@ -44,14 +44,19 @@ install:
...
 	$(IPROG) init.d-xapissl $(DESTDIR)$(LIBEXECDIR)/xapissl
-	$(IPROG) init.d-xapi-domains $(DESTDIR)/etc/rc.d/init.d/xapi-domains
-	$(IPROG) init.d-management-interface $(DESTDIR)/etc/rc.d/init.d/management-interface
-	$(IPROG) init.d-genptoken $(DESTDIR)/etc/rc.d/init.d/genptoken
-	$(IPROG) init.d-save-boot-info $(DESTDIR)/etc/rc.d/init.d/save-boot-info
-	$(IPROG) init.d-attach-static-vdis $(DESTDIR)/etc/rc.d/init.d/attach-static-vdis
-	$(IDATA) cdrommon@.service $(DESTDIR)/usr/lib/systemd/system/cdrommon@.service
+	$(IPROG) init.d-management-interface $(DESTDIR)$(LIBEXECDIR)/legacy-management-interface
+	$(IPROG) init.d-save-boot-info $(DESTDIR)$(LIBEXECDIR)/save-boot-info
+	$(IPROG) init.d-attach-static-vdis $(DESTDIR)$(LIBEXECDIR)/attach-static-vdis
```
Comment (robhoes) [inline scripts/OMakefile]: Why keep these two (save-boot-info and attach-static-vdis)?

File: scripts/OMakefile [hunk 3]
```diff
@@ -127,13 +131,3 @@ install:
 	$(IDATA) gpg/secring.gpg $(DESTDIR)/opt/xensource/gpg/
 	$(IDATA) gpg/pubring.gpg $(DESTDIR)/opt/xensource/gpg/
 	$(IDATA) gpg/trustdb.gpg $(DESTDIR)/opt/xensource/gpg/
-
-sdk-install: install
```
Comment (robhoes) [inline scripts/OMakefile]: Why is this removed?

File: scripts/OMakefile (hunk 2 above)
Comment (robhoes) [inline scripts/OMakefile]: Oh nevermind, I see now that they are moved to libexec.

File: scripts/attach-static-vdis.service [hunk 4]
```diff
@@ -0,0 +1,12 @@
+[Unit]
+Description=Attach statically-configured VDIs to dom0
+Before=xapi.service
+
+[Service]
+Type=oneshot
+RemainAfterExit=yes
+ExecStart=@LIBEXECDIR@/attach-static-vdis start
```
Comment (robhoes) [inline scripts/attach-static-vdis.service]: I think that we need to make it call `attach-static-vdis stop` as well.

File: scripts/OMakefile (hunk 1 above)
Comment (rafalmiel) [inline scripts/OMakefile]: save-boot-info and other oneshot scripts are put into ExecStart entry in the respective service file

**Comment by rafalmiel** [comment]
> As discussed, please add a systemd target for xe-toolstack-restart, rather than redefining the startup order in the script.

This is not straightforward, there seems to be no clean solution for stopping services belonging to the target.
Please see http://unix.stackexchange.com/questions/212438/how-to-stop-all-units-belonging-to-the-same-target

File: scripts/OMakefile (hunk 3 above)
Comment (rafalmiel) [inline scripts/OMakefile]: Looks like `/etc/rc.d/init.d/sdkinit` was removed. It comes from the work made by Simon

File: scripts/OMakefile (hunk 1 above)
Comment (robhoes) [inline scripts/OMakefile]: Yes, but I was talking about `cdrommon@.service` and the new `*.service` files below.

Comment (rafalmiel) [inline scripts/OMakefile]: Ah sorry I missed that, this could probably be `IDATA`

File: scripts/xe-toolstack-restart [hunk 5]
```diff
@@ -24,40 +26,35 @@ fi
...
 
-	# restart other services only if chkconfig said they were enabled
-	chkconfig $svc
+	# restart other services only if systemd said they were enabled
+	systemctl is-enabled $svc >/dev/null 2>&1
 
 	if [ $? -eq 0 ] ; then
 		TO_RESTART="$svc $TO_RESTART"
-		service $svc stop
+		action "systemctl stop $svc" systemctl stop $svc
```
Comment (robhoes) [inline scripts/xe-toolstack-restart]: Can this whole loop be replaced by `system-ctl try-restart $SERVICES`? I suspect that systemd will then take care of the ordering.

> try-restart PATTERN...
> Restart one or more units specified on the command line if the units are running. This does nothing if units are not running.

Comment (robhoes) [inline scripts/xe-toolstack-restart]: Hmm... that is not quite the same as checking `is-enabled` though... :/

Comment (rafalmiel) [inline scripts/xe-toolstack-restart]: Yes that's a good idea, I will give it a try

File: scripts/init.d-xapi [hunk 6]
```diff
@@ -57,30 +49,15 @@ start() {
 	if [ -e ${XAPI_BOOT_TIME_INFO_UPDATED} ]; then
 	    # clear out qemu coredumps/chroot dirs on system boot:
 	    rm -rf /var/xen/qemu/*
-	    "@BINDIR@/xapi" -daemon true ${xapiflags} \
+	    #rm -f ${XAPI_BOOT_TIME_INFO_UPDATED}
+	    exec "/opt/xensource/bin/xapi" -nowatchdog ${xapiflags} \
```
Comment (robhoes) [inline scripts/init.d-xapi]: Can we keep `@BINDIR` please.

File: scripts/init.d-xapi [hunk 7]
```diff
@@ -57,30 +49,15 @@ start() {
...
 		-writereadyfile ${XAPI_STARTUP_COOKIE} -writeinitcomplete ${XAPI_INIT_COMPLETE_COOKIE} -onsystemboot
 		RETVAL=$?
-	    rm -f ${XAPI_BOOT_TIME_INFO_UPDATED}
 	else
-	    "@BINDIR@/xapi" -daemon true ${xapiflags} \
+	    exec "/opt/xensource/bin/xapi" -nowatchdog ${xapiflags} \
```
Comment (robhoes) [inline scripts/init.d-xapi]: `@BINDIR`

File: scripts/init.d-xapi-old [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+start() {
+	systemctl start xapi
+}
+
+stop() {
+	systemctl stop xapi
+}
+
+rhstatus() {
+	status xapi
```
Comment (robhoes) [inline scripts/init.d-xapi-old]: Not `systemctl status xapi`?

File: scripts/init.d-xapi-old [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+restart() {
+	stop
+	start
```
Comment (robhoes) [inline scripts/init.d-xapi-old]: `systemctl restart xapi`?

File: scripts/init.d-xapi-old [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+case "$1" in
+  start)
+	start
```
Comment (robhoes) [inline scripts/init.d-xapi-old]: Since most function are one-liners, might as well have them in-line here.

File: scripts/perfmon.service [hunk 11]
```diff
@@ -0,0 +1,13 @@
+[Unit]
+Description=Performance monitoring/alarm generation daemon
+After=xcp-rrdd.service xapi.service
+Wants=xcp-rrdd.service xapi.service
```
Comment (robhoes) [inline scripts/perfmon.service]: I think that these `After`s and `Wants`s are not strictly needed. Perfmon will most likely just retry.

File: scripts/xapi.service [hunk 12]
```diff
@@ -0,0 +1,21 @@
+[Unit]
+SourcePath=@LIBEXECDIR@/xapi-init
+Description=XenServer server (XAPI)
```
Comment (robhoes) [inline scripts/xapi.service]: Let's call it "XenAPI server (XAPI)"

File: scripts/xapi.service [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+After=network-online.target syslog.target forkexecd.service xcp-networkd.service message-switch.service v6d.service save-boot-info.service attach-static-vdis.service genptoken.service xcp-rrdd.service xapi-storage-script.service xenopsd-xc.service xenstored.service
+Conflicts=shutdown.target
+
+[Service]
+Type=notify
+Restart=on-failure
+TimeoutSec=5min
+IgnoreSIGPIPE=no
+KillMode=process
```
Comment (robhoes) [inline scripts/xapi.service]: So this will be changed to `Type=simple` and I guess we'll keep `Restart=on-failure`, but the other three line can go away?

File: scripts/xapi.service [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+ExecStart=@LIBEXECDIR@/xapi-init start
+ExecStartPost=/usr/bin/touch /var/lock/subsys/xapi
+ExecStartPost=/usr/bin/rm -f /etc/xensource/boot_time_info_updated
```
Comment (robhoes) [inline scripts/xapi.service]: And then this will go too early.

**Comment by robhoes** [comment]
OK, I think this is fine apart from the above.

File: ocaml/xapi/xapi.ml [hunk 15]
```diff
@@ -960,7 +961,7 @@ let server_init() =
 
       "SR scanning", [ Startup.OnlyMaster; Startup.OnThread ], Xapi_sr.scanning_thread;
       "Updating pool cpu_info", [], (fun () -> Create_misc.create_pool_cpuinfo ~__context);
-      "writing init complete", [], (fun () -> Helpers.touch_file !Xapi_globs.init_complete);
+      "writing init complete", [], (fun () -> Helpers.touch_file !Xapi_globs.init_complete; systemd_notify_ready ());
```
Comment (robhoes) [inline ocaml/xapi/xapi.ml]: What is this for?

File: scripts/xapi.service [hunk 16]
```diff
@@ -9,7 +8,7 @@ Type=simple
 Restart=on-failure
 ExecStart=@LIBEXECDIR@/xapi-init start
 ExecStartPost=/usr/bin/touch /var/lock/subsys/xapi
-ExecStartPost=/usr/bin/rm -f /etc/xensource/boot_time_info_updated
+#ExecStartPost=/usr/bin/rm -f /etc/xensource/boot_time_info_updated
```
Comment (robhoes) [inline scripts/xapi.service]: Just remove the line.

File: scripts/init.d-xapi-compat [hunk 17]
```diff
@@ -6,38 +6,22 @@ if [ -f /etc/sysconfig/xapi ]; then
...
-}
-
 case "$1" in
   start)
-	start
+	systemctl start xapi
 	;;
   stop)
+	systemctl stop xapi
 	stop
```
Comment (robhoes) [inline scripts/init.d-xapi-compat]: One more line to remove...

File: ocaml/xapi/xapi.ml (hunk 15 above)
Comment (robhoes) [inline ocaml/xapi/xapi.ml]: I see it goes away in the last commit...

**Comment by robhoes** [comment]
OK, :shipit: after squashing.

File: scripts/attach-static-vdis.service [hunk 18]
```diff
@@ -0,0 +1,13 @@
+[Unit]
+Description=Attach statically-configured VDIs to dom0
+Before=xapi.service
```
Comment (zli) [inline scripts/attach-static-vdis.service]: Not familiar with systemd, would this line be enough to keep the service starting before xapi and also stopping after xapi?

Comment (rafalmiel) [inline scripts/attach-static-vdis.service]: yes:
https://www.freedesktop.org/software/systemd/man/systemd.unit.html#Before=

> Note that when two units with an ordering dependency between them are shut down, the inverse of the start-up order is applied. i.e. if a unit is configured with After= on another unit, the former is stopped before the latter if both are shut down.

**Comment by zli** [comment]
Is there any change on xapi-ssl? 

I'm wondering whether it should be set "Restart=on-failure" like xapi. I had a ticket recently, stunnel daemon existed unexpectedly despite xapi was perfectly healthy. This broke the whole system, which reminded me that it's crucial to keep the stunnel proxy running (at least no harm trying to do so).

File: mpathalert-daemon.service [hunk 19]
```diff
@@ -0,0 +1,9 @@
+[Unit]
+Description=MPathAlert Daemon Service
+
+[Service]
+ExecStart=@BINDIR@/mpathalert -delay 120
+StandardError=null
+
+[Install]
+WantedBy=multi-user.target
```
Comment (robhoes) [inline mpathalert-daemon.service]: We probably don't want this, because mpathalert is not started at start-of-day.

File: scripts/xe-toolstack-restart [hunk 20]
```diff
@@ -30,32 +30,22 @@ SERVICES="perfmon xapi v6d xenopsd xenopsd-xc xenopsd-xenlight xenopsd-simulator
...
 
-for svc in $TO_RESTART ; do
-	# mpathalert-daemon isn't a systemd service so needs to be handled specially
-	if [ $svc == mpathalert-daemon ] ; then 
-		@ETCDIR@/master.d/03-mpathalert-daemon start
-		continue
-	fi
-	action "systemctl start $svc" systemctl start $svc
-done
+systemctl try-restart $TO_RESTART
```
Comment (robhoes) [inline scripts/xe-toolstack-restart]: Use `restart` so that the services get started even if not running.

File: mpathalert-daemon.service (hunk 19 above)
Comment (rafalmiel) [inline mpathalert-daemon.service]: I think the daemon will have to be enabled first to be started during boot

**Comment by thomassa** [comment]
Responding to @zli's thoughts on xapissl and the listening stunnel:
I agree that restarting on failure would be a very good thing to do.
Achieving that might be a bit more complicated though.

At present it is _not_ started through init.d but rather the start-up mechanism is:
- xapi uses the forking-executioner to call `/opt/xensource/libexec/xapissl` with certain arguments.
- that script uses those arguments to write a config file to `/etc/xensource/xapi-ssl.conf`
- then the script starts stunnel, specifying the config file it has just created: `start_daemon ${STUNNEL} ${SSLCONFFILE}`

For a restart-on-failure we could re-use the existing config file, but still I don't know how practical it is to do this within the systemd model.

Bear in mind that xapi also needs to be able to restart the stunnel with new config, by calling `/opt/xensource/libexec/xapissl restart` with appropriate additional arguments.

**Comment by zli** [comment]
@thomassa: I didn't realize `xapissl` is no longer an init.d service (it used to be one, as I can see in 6.2 still).  As it's now some kind of auxiliary  internal service (for xapi), one possibility is to replace the `xapissl` script with some ocaml code and then let xapi launch the `stunnel` directly using `fe` just as a normal program, so that either `xapi` or `fe` could monitor its running status and restart it if necessary (either because of configuration change or an unexpected crash/exit). Now I understand why you suggested to add some kind of "alwaysRun" flag to `fe` :smiley:  .

**Comment by robhoes** [comment]
@zli, @thomassa: Yes, I had thought a while ago, when working in the area, to just translate xapissl into an OCaml module inside xapi, which is responsible for writing the config file, and then have xapi fork stunnel and pass the location of the config file on the command line. The SSL/TLS certificate handling can probably be split off as well. I didn't do it at that time, because it was getting too big a change, but we can reconsider it. We could simply have a systemd service for stunnel without all the special logic of xapissl, and let systemd manage autorestarts etc.
